### ISSUE TITLE
Is it possible to insert mute and solo in the FX Parameters window?

### ISSUE THREAD

**Issue by Lo-lo78** [jcsteh/osara#523]
Hello everyone!
When I try to do an EQ, a compression on the snare and so on, in the FX Parameters window I miss putting the solo on the track I'm working on and then removing the solo to hear how the snare sounds with the whole project.
Unfortunately, it is not possible to mute or solo the track in the FX Parameters window.
I find that when making a mix this can speed up the work a lot, without having to esc or close and then hear and remove the solo to see if the e-q or the compressor has been adjusted well.
If I use a plugin like a channel strip, which contains the entire channel of a mixer, then I find myself, preamp, gate, e-q, compressor and so on; in this case it is very important to be able to have a solo and mute toggle. A channel strip has a lot of parameters and getting in and out of the FX Parameters window can be very confusing and slow.
When making a mix, the speed to compare one sound with another is essential; the brain tends to find a balance, the more time passes between one sound and another and the more difficult it is to understand the differences.
For this, tests (A) and (B) are performed practically instantly.
Greetings!

**Comment by Lo-lo78** [comment]
I wanted to add that when using midi, it happens that note off does not go, in the OSARA FX Parameters window.
I mean the action:
Send all notes off to all MIDI outputs / plug-ins
It always concerns the FX Parameters window, if I work with the midi I have to exit the FX Parameters window to give the note off to the midi.
Let me know if I need to open a new issue.
Thank you!

**Comment by ScottChesworth** [comment]
Nice idea. I'd like to add toggling phase to that list of keystrokes being passed, if possible.
Something that may be worth noting is that I remember when @RDMurray implemented the code to get Control+Space working, he found it difficult to do and said it was kinda ugly. I'm guessing the same would apply to these requests, but perhaps some of his code is reusable.

**Comment by ScottChesworth** [comment]
Also wanted to add, I think a strong justification for this being implemented is to smooth out disparity between the productivity that's currently possible when working with third-party plug-ins in the Params dialog versus what's possible when working with more accessible stock plug-ins in the FX window. Right now, I'm forced to take different approaches when the Params dialog is involved here, often at the expense of efficiency.

**Comment by Lo-lo78** [comment]
Scott, that's right! Sure, third party plugins can be used but such a 
check would make the work a lot smoother.
It would be super!

**Comment by ScottChesworth** [comment]
Hey @Lo-lo78, are you on Windows? Do you know how to take a backup of your current key map, then substitute it for a test version? If so, try this.
I've adjusted the scope for all mappings around track muting, soloing and phase inversion.
If this works, I'll tag a Mac user to duplicate these scopes and test on that platform.
Feel free to shout if you're not comfortable manually switching maps there, I can build OSARA with this test key map included for you to try.

https://www.dropbox.com/s/xeq0uoq7z989dg0/reaper-kb.ini?dl=1

Let me know how it goes. :)

**Comment by ScottChesworth** [comment]
Hey @Lo-lo78, are you on Windows? Do you know how to take a backup of your current key map, then substitute it for a test version? If so, try this.
I've adjusted the scope for all mappings around track muting, soloing and phase inversion. Also included all notes off.
If this works, I'll tag a Mac user to duplicate these scopes and test on that platform.
Feel free to shout if you're not comfortable manually switching maps there, I can build OSARA with this test key map included for you to try.

https://www.dropbox.com/s/tryax7b8qs3jf3r/reaper-kb.ini?dl=1

Let me know how it goes. :)

**Comment by jcsteh** [comment]
Yeah, I have to say the current behaviour in Parameters dialogs really annoys me too. I wonder whether a better approach would be to pass everything through to the main section except the keys that are necessary to work with that dialog (arrow keys, tab, escape, characters, space, maybe more). That's more or less what REAPER does itself in its dialogs. I think I might understand how to implement that... maybe.

**Comment by jcsteh** [comment]
> Something that may be worth noting is that I remember when @RDMurray implemented the code to get Control+Space working,

I don't have any recollection of this and I can't find anything about it in the code. AFAIK, this just always worked. I might be wrong, though? Can you point me at an issue or commit where this got added?

**Comment by Lo-lo78** [comment]
E' ok Scott,

funziona molto bene!

👍

**Comment by Lo-lo78** [comment]
YES! It really is the right logic!

**Comment by jcsteh** [comment]
I've implemented this, but I've run into a nasty snag. Some actions you can now run might take focus away from the Parameters dialog; e.g. if you hit f2 to rename a track or f4 to open the Actions dialog. That leaves the Parameters dialog open, which causes a bunch of unpleasant problems.

The obvious workaround is to make sure that whenever the Parameters dialog loses focus, it automatically closes itself. However, that means that if you alt+tab out of REAPER while the Parameters dialog is open, it will also close itself then. I haven't been able to find a way around that.

Can we live with this?

**Comment by mattymc2** [comment]
I personally could definitely live with this. I can’t see many use cases where I would need to switch out of reaper and back again and land back in the parameters dialogue. I think the use of the function keys absolutely outweighs the minor inconvenience of hitting the P key  again

**Comment by jcsteh** [comment]
@mattymc2, there's a test build in https://github.com/jcsteh/osara/pull/524#issuecomment-846327459 if you want to take it for a spin.

**Comment by Lo-lo78** [comment]
Hi James, everything ok but unfortunately with ctrl + space you activate 
the play / pause action. It would be more comfortable if there was play 
/ stop. I tried it and I was not with the play / pause; I understand 
that you are going to connect to the keymap shortcut but if it is 
possible it would be better to restore the play / stop with ctrl + 
space. Alternatively you could think of using the alt key + a character.

But I think play / stop with ctrl + space.Thank you!

**Comment by mattymc2** [comment]
I won't be able to test it out on windows until Monday but I'll give it a spin on Mac this afternoon. Cheers ￼

**Comment by Lo-lo78** [comment]
Per esempio:

ctrl+spazio: play/stop,

shift+spazio play/pause

**Comment by jcsteh** [comment]
I just fixed the control+space bug alongside another rather nasty bug. Build will be on #524 shortly.

**Comment by Lo-lo78** [comment]
James, I'm trying it and it works very well. THANK YOU VERY MUCH!!!

**Comment by ptorpey** [comment]
Looks like Mute and Solo work in the parameters list using Windows.  Control+space also works to play/pause.  Hitting just space does not seem to do anything.

 

I think that is the expected behavior.

 

Anyway, looks good.

 

--Pete

**Comment by mattymc2** [comment]
Looks like it's working perfectly on Mac. Soloing, unsoloing play, stop, etc. Fantastic work￼

**Comment by Robert-J-H** [comment]
I'm not sure why only the keys with modifiers are sent to the main window. It's for instance not possible to jump between markers since those are just the numbers.
My thinking would be that characters are only caught if the widget demands it.
That is, alphanumeric + symbols + arrows + Delete + backspace + home/end for the filter box. What will control+Arrow do?
For the parameter input field, the alphabetic characters and most symbols (except period or whatever local digits point is used) fall away
The parameter names have to be navigable with alphanumeric characters, so there is not much choice except that space could be passed (unless you've come across names with an initial space). 
The slider itself traps the least amount of keys. Arrows, home and end and page up/down although I would sometimes wish that Shift+arrows would move by less amount (e.g. for frequency sliders).
In summary, in 3 of 4 locations could space be passed through.

**Comment by jcsteh** [comment]
Explicitly choosing what to pass through to the widget turned out to be a lot more complicated than it first seems. It means you have to list a huge number of keys and it's very easy to miss something. For example, it turns out you need control+letters for copy, paste, etc. Control+left/right arrows move by word. Also, capturing symbols is non-trivial, especially if you consider languages other than English.

It should be possible to pass numbers through when not in a text box or combo box if you have an actual need for that.

**Comment by Robert-J-H** [comment]
Other languages are probably irrelevant as foreign characters won't appear in the parameter names, so it is down to ANSI (or ascii although ç,€ and such shouldn't be there) that has to stay in the filter box.
I can see that it is a complicated task not least because of the focus change issue for some actions.
While we're at the parameter box, consider allowing extended filtering the way it is provided in e.g. the FX browser, specifically upper case AND, OR, NOT.
Thus, useless parameters could be removed from the list such as the minus signs in the komplete kontrol parameter list or the 2000 parameters that sneak in for internal MIDI messages in some VST3s.
I will later file an official issue for this. Thanks

**Comment by jcsteh** [comment]
I'd be happy to take a pull request for extended filtering, but writing that kind of parser is not something I anticipate having the time to do.
